### PR TITLE
Mirror clang's link stage on Windows.

### PR DESCRIPTION
**Pull request by dot-asm**
Fixes problem discussed in #811.

### CODE REVIEW

**Comment by ChrisDenton** [comment]
Hm, unsure about the implementation of this. Since we're forcing this library to be included, we need to be careful that it doesn't override anything the user has implemented themselves. clang does this using `-defaultlib:oldnames.lib` to give the library lower priority than non-default libs.

**Comment by dot-asm** [comment]
`cargo:rustc-link-arg=-defaultib:oldnames` then. Deeper dive into `-vvv` logs for a multi-file project revealed that it's also a wrong spot for the link arguments. [As well as --target duplication thing.]

File: src/lib.rs [hunk 1]
```diff
@@ -1172,6 +1172,11 @@ impl Build {
                     atlmfc_lib.display()
                 ));
             }
+
+            if compiler.family == ToolFamily::Clang {
+                // Mirror what clang does at link stage
+                self.print(&format_args!("cargo:rustc-link-arg=-defaultib:oldnames"));
```
Comment (ChrisDenton) [inline src/lib.rs:1178]: Typo
```suggestion
                self.print(&format_args!("cargo:rustc-link-arg=-defaultlib:oldnames"));
```

**Comment by ChrisDenton** [comment]
This could really use a test to make sure it works and isn't accidentally broken in the future.

**Comment by ChrisDenton** [comment]
Also cc @thomcc as using `rustc-link-arg` would require bumping the MSRV to >=1.56 (October 2021).

**Comment by thomcc** [comment]
That MSRV bump doesn't bother me, but I would really like someone to double-check this works. My recollection is that there's some issue with using rustc-link-arg like this.

**Comment by dot-asm** [comment]
> Also cc @thomcc as using `rustc-link-arg` would require bumping the MSRV to >=1.56 (October 2021).

Hmm, that would probably be undesirable... Two points in the context. One should keep in mind that the suggested modification affects only "fringe" use cases, an MSVC target with `$CC` set to `clang`, as opposed to `clang-cl`, which by itself is error-prone in the most general case. With this in mind, secondly, since the directive is ignored by an older version, then it wouldn't pose a problem for users who use the older version.

**Comment by dot-asm** [comment]
Putting together a test case was challenging because one has to mix object files generated by cl and clang in the same executable, with the former smuggling in `oldnames.lib` through the linker directive. `-Zl` prevents it from doing this and thus shifts the burden of linking in its entirety on rustc. You can convince yourself that the test is affirmative by commenting the line that adds `-defaultlib:oldnames` link flag and noting that the cc-test fails to link with "unresolved external symbol strdup" message. Well, this is provided that you have `clang` on the `%PATH%`. Just in case, this is the case on Github Actions, which can be confirmed by noting `test msvcrt_here ... ok` in the log files.

**Comment by dot-asm** [comment]
Re-based just in case.

**Pull request closed without merging**